Azimuthal2D Integration

CSSFrancis opened the pull request:
---
name: Azimuthal2D Integration
about: Changes how the azimuthal integration is handled in the 1 and 2D cases. Uses azimuthal integration for polar reprojection. 

---

**Release Notes**
> new feature
> Summary: Generalizes some of pyFAI's functionality with azimuthal integration.

**What does this PR do? Please describe and/or link to an open issue.**

This pull request simplifies some of the code regarding azimuthal integration in 1 D and "caking" in 2D.  For the most part adds in the ability to preform integration using the units in the axis_manager and assuming a flat Ewald sphere.

This just means creating a generic flat detector and using the  unit="2th_deg" for the integrator.

In addition this allows the user to apply an affine transformation to the data through applying a series of shifts to the x, y positions in a grid.  This could potentially be replaced by a grid like spline for more complicated distortions.

**Describe alternatives you've considered**

There are still a fair number of things to consider regarding giving the use the flexibility to use some of the more complicated integration options available in PyFAI.  Additionally I think that integrating some of the GPU enabled calculations efficiently would be of interest but might require some testing.

**Work in progress?**

- [x] Determine detector set up that creates a 1:1 transcription of the image
- [x] Allow for affine transformations to occur during reprojection
- [x] Allow for unique detector set ups/ transfer units from these set ups
- [x] Transfer axis from axis manager to 2D reprojection correctly
- [x] Transfer axis from axis manager to 1D integration correctly

dnjohnstone commented:
Thanks @CSSFrancis - would be great to extend demo-07 to include this functionality, would also speed up my review.

Basically looks nice - do you know where you're going with the tests? Ideally this method should be tested with an assert on array equality I think.

CSSFrancis commented:
@dnjohnstone This has been a little more difficult than I first realized.  Let me start by just explaining what I think is the way that pyXEM should integrate with pyFAI.

I think there should be two main cases with a couple of options in both.

- [x] Case 1: The user want's to use pyXEM to handle the units
    - [x] subcase 1: The user doesn't care about the Ewald sphere (no wavelength given)
    - [ ] subcase 2: The user doesn't care about the Ewald sphere (wavelength given)
- [x] Case 2: The user want's to use pyFAI to handle the units

Beyond that for both of these cases there should be the ability to pass the following:
- [x] Affine transformation to be applied
- [x] Origin of the beam as pixel coordinates
- [x] Mask to be applied

For the most part all of these have been implemented but I am having a bit of difficulty with Case1-subcase2 due to the lack of standard on what the units are on our part (k, q, s and if they are nm^-1, Angstroms^-1 etc.) I think this is talked about in #456.  

I added this to the metadata as something the user can set ie `signal.unit = "k_nm^-1"` but that might be something that deserves a little more thought going forward.

dnjohnstone commented:
> For the most part all of these have been implemented but I am having a bit of difficulty with Case1-subcase2 due to the lack of standard on what the units are on our part (k, q, s and if they are nm^-1, Angstroms^-1 etc.) I think this is talked about in #456.
> 
> I added this to the metadata as something the user can set ie `signal.unit = "k_nm^-1"` but that might be something that deserves a little more thought going forward.

Thanks for picking this up @CSSFrancis 

Originally we were quite clear on this in pyxem and everything was in "k_A^-1" - which is what I would call the "crystallographic convention" and I think is what most people would expect with electron diffraction. That's how it remains for all the crystallographic mapping and diffraction vectors bits of the code.

Unfortunately that doesn't fit very well with what pyFAI does and also, frustratingly, it just isn't what anybody does in PDF literature (s and q are popular there) - at that point we ended up with different behaviour for the PDF workflow code and the crystallographic mapping workflow code.

The upshot of all that is that I think it's basically not possible to just pick one convention and stick to it in pyxem while also being consistent with what various communities do as "established practice" in literature. I think this is also why pyFAI supports various options.

I think the only option then becomes to track this in metadata as you suggest.

The good news is that I think we can (and should for various reasons) - stick to "k_A^-1" for Diffraction2D signals and then track the units in this azimuthal integration step.

Does that all make sense?

dnjohnstone commented:
@JoonatanL please could you have a look at this branch and let us know if this breaks anything in the PDF pipeline?

dnjohnstone commented:
This is starting to look pretty close and given I haven't yet touched Diffraction2D in #569 - I think it would now make sense to merge this PR before the pixSTEM/pyxem merge and then avoid touching Diffraction2D until after #569 - does that sound ok @CSSFrancis @pc494 ?

What do you think you need to do still @CSSFrancis ?

Do you think you've changed any functionality and could you update demo 7 (github.com/pyxem/pyxem-demos) to make it a bit quicker for me to test this, please?

A couple of things jump out at me:

1. azimuthal_integrate2d - does a polar reprojection? rather than an actual integration?
2. Is it much quicker to not take account of the Ewald sphere? If the speed is similar, why would we not just do it "properly"?
3. I think this might also be the time to get rid of  get_radial_profile() - which is now duplicate functionality, I think we kept it because it was much faster in the past, but I think that might not be true now.

@JoonatanL your input on this one would be much appreciated.

pc494 commented:
@dnjohnstone were you thinking of getting this in for `0.11`? Or as the first PR of `0.12`?

dnjohnstone commented:
> @dnjohnstone were you thinking of getting this in for `0.11`? Or as the first PR of `0.12`?

Ideally last one of 0.11.0 - but I don't think it matters too much?

CSSFrancis commented:
> What do you think you need to do still @CSSFrancis ?
> 
> Do you think you've changed any functionality and could you update demo 7 (github.com/pyxem/pyxem-demos) to make it a bit quicker for me to test this, please?

I have changed around the functionality quite a bit to deal with the case where the units are taken from "pyxem" and added in the ability to use a mask, affine transformation and a changing origin.

@dnjohnstone For the most part I think this is largely finished but before any merges take place my plan is to:

- [ ] Update demo 07
- [ ] Test the speed of the different implementations
- [ ] Write up the different methods available through pyFAI

 I was hoping to get to this over the weekend as a final check that everything is working like it should.  I still haven't totally understood all of the options available with pyFAI. 

In relation to your questions:

> 1. azimuthal_integrate2d - does a polar reprojection? rather than an actual integration?

What it does is dependent on the method that you use and the keyword correctSolidAngle.  There is some discrepancies here that I would like to test a little bit more.  I would assume that if correctSolidAngle=True than overall intensity would be conserved (closer to an actual integration) but that is not always the case.

> 2. Is it much quicker to not take account of the Ewald sphere? If the speed is similar, why would we not just do it "properly"?

That's a good point to make.  The speed is pretty much identical and it's a fairly "hacky" method implementation. The reason I added it was that for the most part that is how electron microscopist are operating now.  I think part of the problem with pyFAI is the large headache associated with setting up some unique geometry which is important for xray experiments but not as important for electron microscopists.  I just wanted a straight forward method that just worked as a starting point and comparison when what they are already doing then let the user decide what else they wanted.

That being said, accounting for the Ewald sphere just means knowing the wavelength, the pixel size and setting the units which should be something everyone doing these experiments should know.

> 3. I think this might also be the time to get rid of  get_radial_profile() - which is now duplicate functionality, I think we kept it because it was much faster in the past, but I think that might not be true now.

I'll need to run some tests to see the speed of the different implementations.  I kept the polar_reprojection method as well which is duplicated here and might remove that it it seems to work fairly similarly.

As a final note it might be worth it in the future to move away from the hyperspy map functionality and use the pyFAI [worker](https://pyfai.readthedocs.io/en/latest/api/pyFAI.html#module-pyFAI.worker) module.  Adding in support for GPU processing might be useful for processing very large datasets.

CSSFrancis commented:
@dnjohnstone  I started working though updating the demo [(my demo)](https://github.com/CSSFrancis/pyxem-demos/blob/master/07%20Azimuthal%20Integration%20Using%20pyFAI%20Detector.ipynb) to show the functionality that I added with the integration.  For the most part thing there should be finished and seem to run rather quickly except in the case where you need to apply a different mask, affine transformation or center to each pattern.

In that case it drops from around 100-200 diffraction patterns a second to around 3 when map_iterate is used. 

The one thing that I am uncertain about is if the correctSolidAngle parameter is actually being applied. I think this is the parameter that corrects based on the actual area integrating which is fairly important for maintaining the total intensity after integration.

I think that the answer is no but I am not sure if that is a bug on pyFAI's side or my own, I assume my own, but from testing what is passed to azimuthal_integrate2d_fast  it seems like I should be passing it correctly. 

I'll have to look over the testing coverage again to make sure every part of the code is hit but everything seemed to be working as I intended for the demo.

dnjohnstone commented:
Ok @CSSFrancis - just a simple thing off the top - the coverage is dipping here because you aren't using the dp_for_origin_variation() fixture for testing [here](https://github.com/pyxem/pyxem/blob/master/pyxem/tests/conftest.py) anymore - was that intentional? Are you still testing that behaviour? If you're using a fixture like that it should probably go in conftest.

If you're not using that fixture anymore though (and are testing the functionality) you can delete it.

CSSFrancis commented:
Ahh, apparently I am not very good at reading the coverage reports, let me take another crack at this this morning.

dnjohnstone commented:
> Ahh, apparently I am not very good at reading the coverage reports, let me take another crack at this this morning.

I'm looking at this: https://coveralls.io/builds/29896034/source?filename=pyxem%2Ftests%2Fconftest.py#L169

Duncan

CSSFrancis commented:
One question about the conftest.py file.  Is that idea that most of the pytest.fixture methods should be in that file if they are used in multiple places?  I might try to clean up some of my tests this morning.  After looking through them I realized there are quite a few places I need to be more specific.  

Another question is how do you feel about using numpy.testing?  I think for array equality 
assert np.allclose(array1, array2, scale) is used most often in the code.

I have been using np.testing.assert_array_almost_equal(array1, array2) but the difference is minimal. I think the argument for using the numpy testing suite is that there is more information given about why a test fails (number of indicies mismatched, maximum difference, etc.) and there are some more specific testing methods.

dnjohnstone commented:
> @dnjohnstone One question about the conftest.py file. Is that idea that most of the pytest.fixture methods should be in that file if they are used in multiple places? I might try to clean up some of my tests this morning. After looking through them I realized there are quite a few places I need to be more specific.

Yeah - if a fixture is used more than one place (or is likely to be in future) it should be in conftest.py not the individual test file.

> Another question is how do you feel about using numpy.testing? I think for array equality
> assert np.allclose(array1, array2, scale) is used most often in the code.
> 
> I have been using np.testing.assert_array_almost_equal(array1, array2) but the difference is minimal. I think the argument for using the numpy testing suite is that there is more information given about why a test fails (number of indicies mismatched, maximum difference, etc.) and there are some more specific testing methods.

I'm fine with using either - if someone can give me a good reason for doing one or the other always then I'd also be happy to adopt a convention on that, but we haven't yet - so whatever you want.

pc494 commented:
Just for context @CSSFrancis, I'm fairly certain that I always use the

`assert np.allclose(a1,a2,rtol)` 

and @dnjohnstone leans `np.testing`

I don't feel that strongly, but I find my syntax particularly simple to read. We certainly haven't agreed a style on this one yet.

CSSFrancis commented:
@dnjohnstone I think that before this is merged we might consider removing the polar_reproject method.  This pull doesn't include something like the Jacobian keyword but it would be fairly easy to add to the azimuthal integration.  I can't find something in pyFAI to handle this but the scaling is fairly easy, I just have to think about if there are any differences due to the Ewald sphere.

dnjohnstone commented:
@pc494 any idea why travis has started erroring on mac?

dnjohnstone commented:
Please could you also have a look to see if demo8 still works ok with this?

CSSFrancis commented:
@dnjohnstone sounds good! 

> * [ ]  methods of Diffraction2D as an API convention should have beam_energy as an argument rather than wavelength (there is a diffsims util to do the conversion). I then also think that if None it should try to read the beam_energy from the metadata and if it's not specified raise an error.
> * [ ]  Related to the above - I think we should insist that people specify the beam_energy, I think you've said that the flat Ewald sphere approach is not much faster and on that basis I think we can just "improve" default workflows by insisting it's specified - I don't think there are very many good reasons for not knowing this!

That sounds like a reasonable change.  I normally use beam_energy but decided to use wavelength to be more inline with the X-ray community and pyFAI.  As for removing the flat Ewald sphere approach that sounds reasonable and will help clean up the method a little.

I think the only problem might be the case where you don't know the pixel size when you are trying to calibrate something.   Reasonably this shouldn't be the case but there might be some edge case where that might occur.  The other case is if you want to compare with old data where you didn't do the correction.  Either way I think is fine

> * [ ]  I think it would be better to store the "unit" in metadata rather than as an attribute because then it would be saved and loaded with the signal by pyxem, I think?

Sorry, that was me just playing around with things.  Currently the unit is saved in the hyperspy metadata.  I just created a getter and setter function which pulls the value from the metadata.  In that way it is more of a "fake" attribute.

Either way works but I think that `signal.metadata.Singal["unit"]="k_nm^-1` is a little more clumsy than `signal.unit="k_nm^-1` if we are going to be using the property in multiple different cases. 

This might bring a bigger issue that while the hyperspy metadata and axes_manger formats are great, they are slightly clumsy (partially because they are so flexible).  Basically most people end up writing some method like the set_experimental_parameters or set_diffraction_calibration in the electron diffraction class.  I think that in those cases [getters and setters](https://www.geeksforgeeks.org/getter-and-setter-in-python/) might make things more reasonable.

Something like `signal.calibration=.09` makes more sense in my mind than `signal.set_diffraction_calibration(.09)`.  



> * [ ]  I think "unit" should take either one of the explicit values or None and if None try to get it from the axes manager (or metadata?) of the Diffraction2D signal.

I agree. I'll make that change


> * [ ]  You've added a significant amount of time to test_diffraction2d.py - obviously we need to make sure everything is covered, but if it's possible to keep the speed down at all, we should. I also noticed that you didn't re-use the old pytest fixtures for azimuthal integration - what was wrong with them? They were nice and small and fairly intuitive?

Let me pair down the tests a bit.  I think I went a little overboard in checking every parameter combination but in reality most of the code is over tested.

> * [ ]  I've made DiffactionVariance classes inherit from Diffraction classes to avoid code duplication - we should write a test to check that when azimuthal integration is applied to DiffractionVariance2D it does return a DiffractionVariance1D - if it doesn't we should fix that.

Yes I can check that.  It should only be a small change.

CSSFrancis commented:
@dnjohnstone I will check demo 8 as well once I make the changes.

dnjohnstone commented:
> Sorry, that was me just playing around with things. Currently the unit is saved in the hyperspy metadata. I just created a getter and setter function which pulls the value from the metadata. In that way it is more of a "fake" attribute.
> 
> Either way works but I think that `signal.metadata.Singal["unit"]="k_nm^-1` is a little more clumsy than `signal.unit="k_nm^-1` if we are going to be using the property in multiple different cases.
> 
> This might bring a bigger issue that while the hyperspy metadata and axes_manger formats are great, they are slightly clumsy (partially because they are so flexible). Basically most people end up writing some method like the set_experimental_parameters or set_diffraction_calibration in the electron diffraction class. I think that in those cases [getters and setters](https://www.geeksforgeeks.org/getter-and-setter-in-python/) might make things more reasonable.
> 
> Something like `signal.calibration=.09` makes more sense in my mind than `signal.set_diffraction_calibration(.09)`.

Ok - I agree with this. So could you keep what you've got and maybe do something equivalent for beam_energy - perhaps this would mean we can remove set_experimental_parameters and create a getter / setter for each of those parameters?

Also I think this should probably now go in CommonDiffraction because it applies equally to Diffraction1D and Diffraction2D? https://github.com/pyxem/pyxem/blob/master/pyxem/signals/common_diffraction.py

dnjohnstone commented:
> I think the only problem might be the case where you don't know the pixel size when you are trying to calibrate something. Reasonably this shouldn't be the case but there might be some edge case where that might occur. The other case is if you want to compare with old data where you didn't do the correction. Either way I think is fine

I think we can deal with that using https://pyfai.readthedocs.io/en/latest/man/pyFAI-calib.html and thinking about how we can incorporate that into the CalibrationGenerator https://github.com/pyxem/pyxem/blob/master/pyxem/generators/calibration_generator.py

My feeling is that we should encourage the "standard workflow" to include determining this with proper calibration experiments. If a user is deciding to do this directly from their own data for some reason then they could probably work round it by setting the pixel size to 1 at first and then recalculating manually.

Does that sound ok to you?

dnjohnstone commented:
@CSSFrancis does that give us a clear plan to tie this off? This is going to be the last thing to go in before the 0.11.0 release now I think!

CSSFrancis commented:
> My feeling is that we should encourage the "standard workflow" to include determining this with proper calibration experiments. If a user is deciding to do this directly from their own data for some reason then they could probably work round it by setting the pixel size to 1 at first and then recalculating manually.

I started looking though this again and realized that the "flat Ewald sphere" case is just the case where you set your units to "2th_rad" or "2th_deg".  In that case the work around for the "flat Ewald sphere" is there but less explicit.

> Ok - I agree with this. So could you keep what you've got and maybe do something equivalent for beam_energy - perhaps this would mean we can remove set_experimental_parameters and create a getter / setter for each of those parameters?
> 
> Also I think this should probably now go in CommonDiffraction because it applies equally to Diffraction1D and Diffraction2D? https://github.com/pyxem/pyxem/blob/master/pyxem/signals/common_diffraction.py
> 
> @pc494 - can you confirm that you're ok with this?

In response to this, how do you feel about overriding the azimuthal_integrate method for the electron_diffraction2d signal?  The problem is that it is kind of strange to check for a beam energy in the diffraction2d signal when it may or may not have a beam energy.  Also for the case of someone doing an X-ray experiment they would need to calculate a fake beam energy.

```
Class ElectronDiffraction2d()

     def get_azimuthal_integrate_2d(self, npt_rad, beam_energy=None, **kwargs)
         if beam_energy is None and self.beam_energy is not None:
              beam_energy = self.beam_energy
        if beam_energy is not None
              wavelength = get_wavelength(self.beam_energy)
        else:
              
         super.azimuthal_integrate_2d(npt_rad=npt_rad, wavelength=wavelength, **kwargs) 
```

I think this gives the flexibility of dealing with both conventions.

dnjohnstone commented:
> In response to this, how do you feel about overriding the azimuthal_integrate method for the electron_diffraction2d signal? The problem is that it is kind of strange to check for a beam energy in the diffraction2d signal when it may or may not have a beam energy. Also for the case of someone doing an X-ray experiment they would need to calculate a fake beam energy.

I think I'd rather not do that for now. Until this comment I'd been thinking about this from the perspective of synchrotron X-ray data, where I think it's reasonably common to talk about the X-ray beam energy in keV?

In any case it wouldn't really be a fake beam energy we're just saying you have to convert the X-ray wavelength to energy, which I think is ok.

Either way - could we revisit this when we get round to adding XrayDiffraction classes in #478 ?

CSSFrancis commented:
Okay that’s reasonable. Is it the same calculation for energy to wavelength for X-rays and electrons?

dnjohnstone commented:
> @dnjohnstone That's reasonable but it isn't the same calculation for energy to wavelength for both X-rays and electrons correct?
> 
> -Edit: I'm not entirely well versed in synchrotron x-ray data but if the calculations are different I would make one last push for keeping wavelength as the base case simply because it is ubiquitous for both X-rays and electrons.

Ok - keep it as you had it then. Keep wavelength in general Diffraction classes and overwrite with beam energy as you have for electrons, we'll revisit when we put in X-ray classes.

CSSFrancis commented:
@dnjohnstone I'm not sure why it is failing on the MacOS build again?  

I think this is done for the most part.  I added a couple of getters and setters to the ElectronDiffraction2D class but kept the original methods just for continuity.  If you want me to take them out that is fine too, it doesn't change overly much. 

I tried to limit the time for some of the tests but let me know if I need to pair things down a little bit more. I think that should be possible.
